firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java

```diff
@@ -235,50 +230,62 @@ private void persistFid(String fid) throws FirebaseInstallationsException {
   }
 
   /**
-   * Registers the FID with FIS servers if FID is in UNREGISTERED state.
+   * Registers the FID with FIS servers if FID is in UNREGISTERED state. Also, refreshes auth token
+   * if expired.
    *
    * <p>Updates FID registration status to PENDING to avoid multiple network calls to FIS Servers.
    */
-  private Task<String> registerFidIfNecessary(
-      PersistedFidEntry persistedFidEntry, AwaitListener listener) {
-    String fid = persistedFidEntry.getFirebaseInstallationId();
-
-    // Check if the fid is unregistered
-    if (persistedFidEntry.getRegistrationStatus() == RegistrationStatus.UNREGISTERED) {
-      updatePersistedFidWithPendingStatus(fid);
-      executeFidRegistration(persistedFidEntry, listener);
-    } else {
-      updateAwaitListenerIfRegisteredFid(persistedFidEntry, listener);
-    }
+  private Task<String> registerFidIfNecessary(AwaitListener listener) {
+    synchronized (persistedFid) {
```

Comment:
if you're going to start synchronizing these instead of piping the values through a Task chain then you should add @GuardedBy annotations http://errorprone.info/bugpattern/GuardedBy

Reply:
So, I am locking on persistedFid and not persistedFidEntry( was piping it earlier on this value through task).

Also, I cannot synchronize on persistedFid throughout this class.

Reply:
Yes, my point was that synchronization isn't necessarily required if you're just piping them through a Task call stack.

Why are we synchronizing on it here if we can't do it consistently? What benefit does it provide here that it doesn't provide elsewhere? Can we document that more explicitly?

Reply:
PENDING state was introduced to avoid multiple network calls for getId(). Registration status changes from UNREGISTERED to PENDING before triggering the network call.

Need for synchronization: Reading the existing registration status and updating it to PENDING should be atomic. That's the block i tried to synchronize. 

I think piping wont solve the atomicity issue.